dynamic_programming/longest_common_string.cpp

```diff
@@ -1,53 +1,142 @@
+    return {TestCase("", "", 0),
+            TestCase("ab", "ab", 2),
+            TestCase("ab", "ba", 1),
+            TestCase("", "xyz", 0),
+            TestCase("abcde", "ace", 3),
+            TestCase("BADANA", "ANADA", 3),
+            TestCase("BADANA", "CANADAS", 3),
+            TestCase("a1a234a5aaaa6", "A1AAAA234AAA56AAAAA", 6),
+            TestCase("123x", "123", 3),
+            TestCase("12x3x", "123", 3),
+            TestCase("1x2x3x", "123", 3),
+            TestCase("x1x2x3x", "123", 3),
+            TestCase("x12x3x", "123", 3)};
+}
+
+/**
+ * @brief checks the function longest_common_string_length agains example data
+ * @param test_cases list of test cases
+ * @tparam type representing a list of test cases
+ */
+template <typename TestCases>
+void test_longest_common_string_length(const TestCases& test_cases) {
+    for (const auto& cur_tc : test_cases) {
+        assert(longest_common_string_length(cur_tc.string_a, cur_tc.string_b) ==
+               cur_tc.common_string_len);
+    }
+}
+
+/**
+ * @brief checks the function longest_common_string_length returns the same
+ * result when its argument are flipped
+ * @param test_cases list of test cases
+ * @tparam type representing a list of test cases
+ */
+template <typename TestCases>
+void test_longest_common_string_length_is_symmetric(
+    const TestCases& test_cases) {
+    for (const auto& cur_tc : test_cases) {
+        assert(longest_common_string_length(cur_tc.string_b, cur_tc.string_a) ==
+               cur_tc.common_string_len);
+    }
+}
+
+/**
+ * @brief reverses a given string
+ * @param in_str input string
+ * @return the string in which the characters appear in the reversed order as in
+ * in_str
+ */
+std::string reverse_str(const std::string& in_str) {
+    return {in_str.rbegin(), in_str.rend()};
+}
 
-    cout << "longest string size = " << ma /*<<" "<<indi<<" "<<indj*/ << "\n";
-    for (i = indi - 3; i < indi; i++) cout << str1[i];
-    cout << "\n";
+/**
+ * @brief checks the function longest_common_string_length returns the same
+ * result when its inputs are reversed
+ * @param test_cases list of test cases
+ * @tparam type representing a list of test cases
+ */
+template <typename TestCases>
+void test_longest_common_string_length_for_reversed_inputs(
+    const TestCases& test_cases) {
+    for (const auto& cur_tc : test_cases) {
+        assert(longest_common_string_length(reverse_str(cur_tc.string_a),
+                                            reverse_str(cur_tc.string_b)) ==
+               cur_tc.common_string_len);
+    }
+}
+
+/**
+ * @brief runs all tests for longest_common_string_length funcion
+ */
+void tests() {
+    const auto test_cases = get_test_cases();
+    assert(test_cases.size() > 0);
+    test_longest_common_string_length(test_cases);
+    test_longest_common_string_length_is_symmetric(test_cases);
+    test_longest_common_string_length_for_reversed_inputs(test_cases);
+}
+
+int main() {
+    tests();
```

Comment:
```suggestion
    tests(); // Run self-test implementations
```

Reply:
I am not sure if adding this comment will make the code any more readable.

Reply:
The documention asks it so we must keep the tradition alive.

Reply:
It is written [there](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#documentation):
_Make sure you put useful comments in your code. Do not comment on obvious things._